### PR TITLE
periph/flashpage: deprecate *_free functions

### PR DESCRIPTION
**Pull request by Ollrogge**
### Contribution description
The functions `flashpage_first_free` and `flashpage_last_free` added with #16972 were introduced to avoid overwriting and therefore corrupting firmware / bootloaders when writing to flash memory. These functions have the flaw that if multiple modules simultaneously use them they will not be aware of each other. E.g. `flashpage_first_free` will return the same value for all modules and can therefore lead to the modules corrupting each others data by writing to the same flashpage.

With #17436 being merged a new and better mechanism has been introduced that achieves the same goal as #16972 but without the flaws. It enables the reservation of flash memory at compile time and ensures that no data is corrupted. Furthermore multiple modules can use this mechanism simultaneously. Based on this the `flashpage_*_free` functions are no longer needed and can be deprecated.

I have never deprecated anything, therefore I need some help of what I need to to. Thanks !

### CODE REVIEW

**Comment by Ollrogge** [comment]
@chrysn could you give me some pointers here ?

File: drivers/include/periph/flashpage.h [hunk 1]
```diff
@@ -238,6 +238,7 @@ void flashpage_erase(unsigned page);
 
 /**
  * @brief Get number of first free flashpage
+ * @deprecated will be removed ...
```
Comment (chrysn) [inline drivers/include/periph/flashpage.h:241]:
```suggestion
 * @deprecated Use @ref FLASH_WRITABLE_INIT instead, which is usable in modules as well as
 * applications. The function will be removed after the 2022.04 release.
```

**Comment by chrysn** [comment]
Deprecation notes should always tell why something is deprecated, and what to do instead (as in the suggestion; the "why" doesn't need to be fully explicit). Unless something was introduced only after the latest release anyway, we generally keep a deprecation period of one release, meaning that (if this gets merged fast, which it will) the deprecation notice will stay in until the 2022.04 release, after which a follow-up PR can reap the deprecation and remove the code. (It doesn't need to happen immediately, nobody will complain if it stays longer, but at least there should be one release in which a stable-to-stable upgrader can see that this is deprecated now, migrate inside that release, and get code that'll work on the next release as well).

**Comment by Ollrogge** [comment]
Added your suggestions. Thanks for the help and explanation of the process :)

**Review by chrysn — APPROVED** [review]

**Comment by Ollrogge** [comment]
I guess @ref doesn't work for macros so I removed it

**Comment by chrysn** [comment]
It does, but the documentation of FLASH_WRITABLE_INIT is not built because it is conditional on a module; the condition should be [similar to here](https://github.com/RIOT-OS/RIOT/blob/237f0ab6191f5b3d8ee5de72b856705d5c8cf40b/drivers/include/periph/flashpage.h#L141), note that `#ifdef` needs to become `#if defined()` then. Could you restore the @ref version and activate the documentation in a separate commit?

**Comment by chrysn** [comment]
Renders fine now. Thank you! Should be auto-merged soon.

**Pull request merged**